docs/serialisation/protobuf.md

```diff
@@ -1,86 +1,172 @@
+| string               | N/A           | google.protobuf.StringValue     |
+| integer, byte, short | N/A           | google.protobuf.Int32Value      |
+| integer, byte, short | FIXED         | alloy.protobuf.FixedInt32Value  |
+| integer, byte, short | FIXED_SIGNED  | alloy.protobuf.SFixedInt32Value |
+| integer, byte, short | SIGNED        | alloy.protobuf.SInt32Value      |
+| integer, byte, short | UNSIGNED      | google.protobuf.UInt32Value     |
+| long                 | N/A           | google.protobuf.SInt64Value     |
+| long                 | FIXED         | alloy.protobuf.Fixed64Value     |
+| long                 | FIXED_SIGNED  | alloy.protobuf.SFixed64Value    |
+| long                 | SIGNED        | alloy.protobuf.SInt64Value      |
+| long                 | UNSIGNED      | google.protobuf.UInt64Value     |
+
+#### alloy.proto#protoNumType
+
+Integer and Long shapes can be annotated with the `@alloy.proto#protoNumType` in order to signal what encoding should be used during protobuf serialisation.
+
+- SIGNED
+- UNSIGNED
+- FIXED
+- FIXED_SIGNED
+
+See [here](https://protobuf.dev/programming-guides/proto3/#scalar) for documentation about these encodings.
+
 
 #### UUIDs
 
-`alloy` gives special meaning to the `alloy#UUID` shape in the context of protobuf : `UUIDs` are supposed to be encoding by means of a message containing two long values, which is more optimal than serialising a string.
+By default, string shapes annotated with `@uuidFormat` are serialised as protobuf strings. However, a `alloy.proto#protoCompactUUID` trait is provided, which signals that the serialised form should be a message containing two int64 values :
 
 Smithy:
 
 ```smithy
+use alloy#uuidFormat
+use alloy.proto#protoCompactUUID
+
+@protoCompactUUID
+@uuidFormat
+string MyUUID
+
 structure Foo {
-  @required
   uuid : alloy#UUID
 }
-
 ```
 
 Proto:
+
 ```proto
-message UUID {
+message MyUUID {
   int64 upper_bits = 1;
   int64 lower_bits = 2;
 }
 
 message Foo {
-  uuid: UUID
+  uuid: MyUUID
 }
 ```
 
-#### Aggregate Types
+### Documents
 
-##### Structure
+Documents should be serialised using a protobuf message equivalent to the one below :
+
+```proto
+message DNull {
+}
+
+message DArray {
+  repeated Document values = 1;
+}
+
+message DObject {
+  map<string, Document> values = 1;
+}
+
+message Document {
+  oneof value {
+    DNull dNull = 1;
+    bool dBoolean = 2;
+    double dNumber = 3;
+    string dString = 4;
+    DArray dArray = 5;
+    DObject dObject = 6;
+  }
+}
+```
+
+For convenience, alloy provides this proto definition in a jar publised to maven-central `alloy-protobuf`
+
+### Aggregate Types
+
+Structure, unions and (string) enumerations follow the rule that unless specified otherwise, the "proto indexes" (or "field numbers") assigned
+to their members should be monotonically increase. The first member receives the index `1` in the case of structures and unions, whilst it receives the index `0` in the case of enumerations.
```

Comment:
For structures at least, we don't enforce that the first index is 1 or that they increase. As far as I can tell, the only validation is that there are no duplicate indices. I'm not sure about unions

Reply:
@lewisjkl, in this paragraph I'm talking about what happens in the absence of `protoIndex`. Granted it's poorly phrased, I'll work on it.

Reply:
[b3faf4f](https://github.com/disneystreaming/alloy/pull/131/commits/b3faf4f9dcc5a2228d72385c8641334eef4bbc32)

Reply:
Ah I see, and yeah that revised version sounds good to me. Thanks